File: 2090-k-radius-subarray-averages/2090-k-radius-subarray-averages.py
class Solution:
    def getAverages(self, nums: List[int], k: int) -> List[int]:
        if k == 0:
            return nums
        if len(nums) < 2*k + 1:
            return [-1]*len(nums)
        prefix = [nums[0]]
        for i in range(1, len(nums)):
            prefix.append(nums[i] + prefix[len(prefix) - 1])
        
        avgs = [-1] * k # since its 0th-indexed
        for i in range(k, len(nums) - k):
            if i-k != 0:
                val = (prefix[i+k] - prefix[i-k-1]) // (2 * k +1)
            else:
                val = prefix[i+k] // (2 * k +1)
            avgs.append(val)
        
        for j in range(i+1, len(nums)):
            avgs.append(-1)
        return avgs
        

File: 2090-k-radius-subarray-averages/2090-k-radius-subarray-averages.java
class Solution {
    public int[] getAverages(int[] nums, int k) {
        // When a single element is considered then its average will be the number itself only.
        if (k == 0) {
            return nums;
        }

        int windowSize = 2 * k + 1;
        int n = nums.length;
        int[] averages = new int[n];
        Arrays.fill(averages, -1);

        // Any index will not have 'k' elements in it's left and right.
        if (windowSize > n) {
            return averages;
        }

        // Generate 'prefix' array for 'nums'.
        // 'prefix[i + 1]' will be sum of all elements of 'nums' from index '0' to 'i'.
        long[] prefix = new long[n + 1];
        for (int i = 0; i < n; ++i) {
            prefix[i + 1] = prefix[i] + nums[i];
        }

        // We iterate only on those indices which have atleast 'k' elements in their left and right.
        // i.e. indices from 'k' to 'n - k'
        for (int i = k; i < (n - k); ++i) {
            int leftBound = i - k, rightBound = i + k;
            long subArraySum = prefix[rightBound + 1] - prefix[leftBound];
            int average = (int) (subArraySum / windowSize);
            averages[i] = average;
        }

        return averages;
    }
}